### ISSUE TITLE
[1.8.1] Mongo database ops blocking the Node event loop, unhandled by Fibers

### ISSUE THREAD

**Issue by chrisbobbe** [meteor/meteor#10677]
First of all, thank you to the Meteor maintainers for your timely response to our previous issue (#10512)! Same as with that one, I'm happy to contribute as much as I can with my limited knowledge of Meteor internals.

## The Problem

Mongo database operations are blocking the Node event loop in a way that isn't (or can't be) handled by Fibers, so all users of my app are affected. Calling this.unblock() doesn't help. Increasing UV_THREADPOOL_SIZE doesn't help.

## Reproduction

* GitHub: https://github.com/chrisbobbe/issue-10677-reproduction/tree/master
* Reproduction provided in 1.8.1, but also observed in 1.9-alpha.9 (both tested on macOS 10.14.5).
* Also observed on a production server with Meteor 1.8.1-rc.1 on Ubuntu 16.04.4 LTS (GNU/Linux 4.4.0-116-generic x86_64).
* I have not tested with a Meteor version earlier than 1.8.1.

## Discovery:

I wanted to reduce the completion time of a Meteor method that does a lot of similar, independent Mongo operations: similar, meaning I can map over an array to produce each query, and independent, meaning a given operation does not depend on the result of the previous operation. My solution was to use a Promise.all, passing an array of promises returned by calls to (for example) .insert() on my Mongo.Collection#rawCollection() (https://docs.meteor.com/api/collections.html#Mongo-Collection-rawCollection)<sup>1</sup>:

```
await Promise.all(
    myArrayOfObjects.map(anObject => {
        const { foo, bar } = anObject;
        return MyCollection.rawCollection().insert({ foo, bar }); // returns a Promise
    })
);
```

instead of

```
for (let index = 0; index < myArrayOfObjects.length; index++) {
    const anObject = myArrayOfObjects[index];
    const { foo, bar } = anObject;
    await MyCollection.rawCollection().insert({ foo, bar });
}
```

The first code example (using Promise.all) is, predictably, faster (29.8 seconds for 5000 inserts, vs. 71.3 seconds for 5000 inserts). The improvement is especially noticeable with a setup where the network path between the application and the database causes higher latency for each individual operation, because it doesn't have to wait for acknowledgement from the database of one op's completion before starting the next. So, at least *some* of the time spent on each query is non-blocking, asynchronous I/O. So far so good for my optimization.

## Expected Behavior:

I expect all (or very nearly all) of the time spent on each query to be non-blocking I/O.

## Observed Behavior:

However, for all (or very nearly all) of those 29.8 seconds, the entire Node event loop is blocked and the server hangs for all users. Printing '...' at a 500ms interval is paused until the event loop unblocks, and other Meteor methods have to wait for the unblocking before they can start.

## Further Concern:

Maybe using a Promise.all() isn't typically necessary (I'm open to other ideas), and it's true that the continous block can be avoided by running all operations sequentially, even though this stretches out the wait time for methods running this many operations. HOWEVER, even without deep knowledge of Meteor and Mongo internals, doesn't it seem plausible that 5000 operations that run in a Promise.all(), and 5000 operations that run sequentially, are actually blocking the Node event loop for exactly the same cumulative duration -- just with the difference that it's clearly visible when using a Promise.all() because the blocking is continuous rather than regularly spaced across time? Can we be sure that running even one Mongo operation is reasonably free from blocking code? Or could the problem have only surfaced now because the issue was able to hide in the sequential case, due to some "breathing room" of non-blocking I/O also included in each operation? I'm getting reports that my app is still slow when I run everything sequentially.

------

<sup>1</sup> This can also be achieved without using .rawCollection(), but my app already uses .rawCollection() for multi-document transactions, so this is where the problem surfaced for me. I see the same blocking behavior with the following code:

```
await Promise.all(
    myArrayOfObjects.map(anObject => {
        const { foo, bar } = anObject;
	return new Promise((resolve, reject) => MyCollection.insert({ foo, bar }, (e, r) => {
            if (e) {
                reject(e);
            } else {
                resolve(r);
            }
        }));
    })
);
```

**Comment by chrisbobbe** [comment]
Running my reproduction with Meteor 1.8.1, I see the following stack trace provided by blocked-at, an NPM package for tracking Node event loop blocks: https://www.npmjs.com/package/blocked-at

<img width="1638" alt="1 8 1" src="https://user-images.githubusercontent.com/22248748/63372748-5dca9880-c34c-11e9-8852-78bf970bb38c.png">

**Comment by chrisbobbe** [comment]
I ran it once on Meteor 1.9-alpha.9, the most current release at this time; again for testPromiseDotAll:
<img width="1647" alt="1 9-alpha 9" src="https://user-images.githubusercontent.com/22248748/63372816-82bf0b80-c34c-11e9-92f6-9c52a325c9c9.png">

**Comment by evolross** [comment]
Interesting.  As a side note for you use-case and application performance - have you considered using one bulk insert versus a loop of individual inserts?

**Comment by chrisbobbe** [comment]
I haven't, really; my example here is a simplification of my use cases, which are many and varied (including not just inserts, but updates and finds also). I can roughly characterize each of them if I have time, but my objection to using Bulk() (https://docs.mongodb.com/manual/reference/method/Bulk/#Bulk) just about anywhere takes the following form:

For each item in a given array `myArray` (of documents, of results from an external API, etc.), I need to execute a function `processElement()`. `processElement()` is written and editable by me, but it contains one **or more** tasks. One such task might be a `.insert()` operation; or `.update()`; or a reusable function `reusableDoOps()` that does multiple operations, or some restructuring of data to be exposed on an external API. It might (here goes) be a `.find()` that returns an array `myOtherArray`, each item of which needs to be passed to `processElement2()`, where `processElement2()` works the same way as `processElement()`. While it may technically be possible to dissect each of these tasks and make versions that use `Bulk()` (as far as "task" refers to a Mongo write operation, which is limiting), I find it's much more manageable to work with `Promise`s and `Promise.all()`, especially for cases where `reusableDoOps()` was originally designed to be run on a single item, not each item in a (large) array, and is readable/easy to maintain according to that design.

**Comment by corporatepiyush** [comment]
It's due to Fibers module it seems. I have done similar stuff with plain async/await. It queues up and executes gracefully. 

I might be wrong but I think the very Fibers module was written in c++ with black magic of asynchronicity for the same purpose of async/await. Its time to move on to native features of NodeJS instead of using this add-on

**Comment by corporatepiyush** [comment]
Faster async functions and promises -> https://v8.dev/blog/fast-async

**Comment by mitar** [comment]
Could it be that the issue here is that you are using only one connection to Mongo? Because you are starting all queries from same fiber. I do not think this really blocks other users in other fibers using other connections?

**Comment by chrisbobbe** [comment]
As described, calling this.unblock() -- ensuring that the queries are started from different fibers, even if initiated by the same user -- does not help. The same problem is observed when you go to the trouble of running the queries from different users.

**Comment by chrisbobbe** [comment]
@corporatepiyush I'm hoping not to have to wait for a major overhaul of Meteor's dependency on fibers; this sounds like it would involve a longer timeline than would be desired for an urgent hotfix. Right now, I'm hoping a Meteor maintainer will triage this issue so I know if it's a quick fix for someone who knows more about Meteor's internals than I do; otherwise, I'll dig deeper at the risk of spending a lot more time than might be necessary.

**Comment by smeijer** [comment]
@chrisbobbe, did you find any way to resolve/work around the issues? I'm having similar problems, and it doesn't take a lot of queries to start being noticeable.

I'm using `rawCollection()` everywhere so that I can use the promise syntax, and so that I'm not limited by the methods that Meteor has wrapped with fibers and not (aggregate for example).

I'm having the same pattern that sounds similar to yours. Some follow up queries that depend on each other, and sometimes a `Promise.all` to await a few (2 / 3 in my case) at the same time.

Unfortunately, the app is starting to get slow. `blocked-at` reports locked event loops on the lines where I call a mongo method (like `await db.coll.find({}).toArray()`.

My data flow is something like:

```
graphql resolver 
  -> apollo datasoure   // db.col = rawCollection()
    -> await db.colX.findOne() // let's say auth check
    -> await db.colY.find({}).toArray()
    -> await Promise.all([
         db.colY.update(...),
         db.colZ.insert(...),
       ]);
  -> return await db.colY.find({})  // return updated doc from resolver
```

Nested graphql resolvers (apollo) are very likely to result in multiple db hits. And with compartmentalizing our frontend, it's also very likely that the client requests multiple queries at the same time. Even if mongo is blocking the event loop for "just 20ms", this is quickly adding up.

**Comment by smeijer** [comment]
Worth mentioning, I'm also using `agenda` to manage some cronjobs. As those queries run quite often and are also blocking, it has a direct impact on the entire stack.

**Here to stack traces from blocking things:**

Something in `meteor` itself?

![image](https://user-images.githubusercontent.com/1196524/70346950-c12af880-185f-11ea-9c66-2a36aa23e17d.png)

A `graphql request`, that was blocking in the lib, and not my own code

![image](https://user-images.githubusercontent.com/1196524/70346985-d4d65f00-185f-11ea-95c5-a90fcdb2fe02.png)

`agenda` / `mongo`

![image](https://user-images.githubusercontent.com/1196524/70347021-ecade300-185f-11ea-8012-69c8950883dd.png)

**Comment by gbhrdt** [comment]
Actually I think it's pretty sad that since a few weeks the stale bot closes very important bugs like this one and many others so they are getting no more attention.

**Comment by smeijer** [comment]
I think the bad thing about stalebot is that it promotes fake activity like this. Commenting "+1" is seen as bad behavior because it adds noise to already busy developers. But not doing so is a call for stale to close the issue.

**Comment by mitar** [comment]
Stale bot should also respect reactions to the issues for issue activity. Not sure if there is GitHub API to get timestamps of when somebody made a reaction to the issue, but bot could just record the counts and then next time it goes around checks if counts increased since the last time, if so, do not mark it stale.

**Comment by sebakerckhof** [comment]
@chrisbobbe have you tried running similar code in plain node? 
If can imagine this just fills up the event loop too much. And timers like setInterval are typically handled in low priority.

**Comment by chrisbobbe** [comment]
@sebakerckhof I believe the reproduction I provided is minimal enough to show the symptoms described in this issue. My use of setInterval there is only to make the blocking obvious, to help with debugging, as a substitute for any other action the server might take — and that includes other database operations, which is why I filed this issue. I have not found any action that the event loop considers high priority enough to take during the blocking period.

Just to clarify, are you proposing (a) a workaround, (b) a debugging strategy, or (c) that Meteor is not responsible for keeping the event loop unblocked? Do you have something specific in mind for "similar code in plain node"?

**Comment by sebakerckhof** [comment]
b -> debugging strategy, by comparing behavior in plain node.
Something comparable would be something that connects to mongodb and does the same amount of queries in the same timespan. Because if you use a rawCollection and do all of the queries in one go, there shouldn't be a lot of Meteor specific code get in your way, except maybe the oplog driver that has to parse the entries.

Another thing that could help is a v8 profile. If you run meteor with the --inspect flag, you should be able to take a CPU profile using the chrome devtools. Be sure to do this without the `blocked-at`, since that blurs the result by being using a lot of CPU itself (ref. https://github.com/meteor/meteor/issues/9828#issuecomment-392603489 )

**Comment by sebakerckhof** [comment]
Okay, I went ahead and ran the demo application. Here is a CPU profile:
[CPU-20200116T203529.zip](https://github.com/meteor/meteor/files/4073390/CPU-20200116T203529.zip)

You can load this in chrome devtools.
As you can see, this doesn't look to be any Meteor specific thing. All of the time is spend by the node-mongodb driver. It's just the time it takes when you do a thousand calls to the node-mongodb driver, which will do a bit of synchronous work on each message, so the whole thing becomes a rather long piece of synchronous work.
I'm fairly certain you would so this outside of Meteor as well, with a script as described above.

Possible solutions are to try to use the batched methods of the node-mongodb driver, or split up your thousand calls into multiple promises, so the event loop isn't blocked for so long.

**Comment by chrisbobbe** [comment]
I've moved permanently to other, non-Meteor projects, and I'm afraid I can't currently commit much more time to this. Thanks for not immediately closing this issue, as I'm not sure it's been nailed down conclusively.

In particular: The ratio of time spent by Mongo to time spent by Meteor doesn't really point to the source of the problem. We're looking at a very large number of very small tasks, not a single monolithic task that blocks the event loop by necessity. Apparently each task does involve a nontrivial amount of synchronous work (and why? -- this is i/o, not heavy computation), but it doesn't follow from that that the event loop can't continue between the tasks.

**Comment by chrisbobbe** [comment]
I agree, though, that trying this outside of Meteor is the best next step.

**Comment by sebakerckhof** [comment]
@chrisbobbe Ok, I went ahead and did the following:
1. Cloned your repo here: https://github.com/sebakerckhof/issue-10677-reproduction
2. Simplified things a bit, so only the rawCollection loop is there
3. Added a plain-nodejs version:
https://github.com/sebakerckhof/issue-10677-reproduction/blob/master/vanilla.js

The results are the following:
1. The plain nodejs version has similar performance like the Meteor one.
2. Hence, also confirmed by the CPU profiles, all time is spend by the calls to the mongodb library.

The good news is, this seem to be solved in mongodb npm library 3.3.2 (currently meteor is on 3.3.0).
And more good news is that a PR to upgrade mongodb npm to 3.4.0 has just been merged a couple of days ago:
https://github.com/meteor/meteor/tree/devel/packages/npm-mongo

TL;DR: The issue is in the mongodb library, which is fixed in 3.3.2, which will soon be part of Meteor.

**Comment by mitar** [comment]
Unstale.

**Comment by jasonnathan** [comment]
Haha!.. Unstale...

**Comment by sebakerckhof** [comment]
Can you guys try the 1.10 rc, this should solve the mentioned problems as per https://github.com/meteor/meteor/issues/10677#issuecomment-575672953

**Comment by smeijer** [comment]
Can someone who's running 1.10 confirm that this has indeed been fixed?

**Comment by gbhrdt** [comment]
With Meteor 1.8.1 we had regular CPU Spikes to 100% lasting forever and resulting in a crash.
I can confirm that we do not have these anymore after upgrading from 1.8.1 to 1.10. I'm not sure if it's because of this fix or just because of the updated node version in 1.10.

**Comment by Floating-action-buttonee** [comment]
I faced a comparable issue and found a suitable resolution https://nodejser.space/questions/a-guide-on-scheduling-automatic-data-insertion-in-mongoosejs
